### DISCUSSION TITLE
Clarify the podman client <-> server compatibility rules

### DISCUSSION THREAD

**Discussion by Romain-Geissler-1A** [containers/podman#14277] [Q&A] [answered]
Hi,

I have a question wrt the compatibility rules between podman client and podman server, especially with the recent deployment of podman 4.x in fedora/RHEL 8.6/RHEL 9.

My own experience seem to infer that:
 - A podman client 3.x will accept to communicate with a podman 3.y server, no matter if x > y or x < y.
 - A podman client 4.x will accept to communicate with a podman 4.y server, no matter if x > y or x < y.
 - A podman client 3.x will accept to communicate with a podman 4.y server.
 - But a podman client 4.x will refuse to communicate with a podman 3.y server.

Are these indeed the right compatibility rules ?

If yes, I have a question. How are we meant to manage transition periods when you bump the major podman digit ?

Context: we are in a corporate environment. Corporate environment implies we use entreprise distros, so RHEL in our case for the host machines. RHEL always trails behind fedora in term of podman version. On our new RHEL development machines, basically nothing except podman is installed, and we chose to run all our development actions inside containers that each developer is free to provide/manage, and basically install any needed software in a self-service mode (it's more or less the idea of ubi-toolbox). The thing is, we don't always necessarily use RHEL/ubi images inside our containers, we sometimes use Fedora (since RHEL/UBI is quite limited in term of package availability, when you compare to what can be installed by default on a Fedora image). We also want to be able to run containers side-by-side from inside the container, so we usually bind mount the podman socket inside the container. When the container is a fedora container, the client podman is usually more recent than the server podman. And during transition from podman 3.x to 4.x, this yields error.

Note: I have forced the migration to RHEL 8.6 in my company, it will be done on all our RHEL development machine this week-end exactly to fix this kind of problems. Technically we had this issue since couple of days. Fedora 35 has podman 3, while Fedora 36 released around the same time RHEL 8.6 has podman 4. Yet, upgrading a container image (fedora) goes usually much faster than upgrade the host OS (RHEL). So, would it make sense to increase a bit the compatibility between podman client and server a bit to allow these interactions between RHEL hosts and Fedora containers ?

Cheers,
Romain

**Comment by rhatdan** [comment, marked as answer]
I would suggest we have to be better and fix the last problem.  Where Podman client 5.x will communicate with Podman 4.x servers.  Basically we need to negotiate the connection and make sure this works, in the future.

**Reply by Romain-Geissler-1A** [reply]
Thanks Dan for confirming this, so normally it's the last time we shall encounter this client <-> server incompatibility when using rather up to date RHEL/CentOS stream/Fedora ;)

**Reply by rhatdan** [reply]
Yes, I would argue that this is a bug for next time we change the API Version, that we need backwards compatibility on the server.
